### PR TITLE
Use user-supplied descriptor pools in FileDescriptor

### PR DESCRIPTION
**Pull request by belvedere-trading-user**
FileDescriptor's **new** accepted a pool argument but did not
honor it. When using the C++ backend implementation, this causes
problems for those who need hot-reloading of Protobuf definitions.

This issue is addressed by honoring the pool argument if it is
supplied.

### CODE REVIEW

**Comment by belvedere-trading** [comment]
In our internal testing, this change seems to allow us to reload our Protobuf files without causing issues. Please let me know if there are any potential issues that you foresee arising from this or if there's further testing you'd like us to complete in order to get this merged.

Thanks!

**Comment by belvedere-trading** [comment]
We've signed the CLA.

**Comment by xfxyjwf** [comment]
ok to test

File: python/google/protobuf/descriptor.py [hunk 1]
```diff
@@ -832,8 +832,11 @@ def __init__(self, name, package, options=None, serialized_pb=None,
     super(FileDescriptor, self).__init__(options, 'FileOptions')
 
     if pool is None:
-      from google.protobuf import descriptor_pool
-      pool = descriptor_pool.Default()
+      if api_implementation.Type() == 'cpp':
+        pool = _message.default_pool
```
Comment (amauryfa) [inline python/google/protobuf/descriptor.py]: Is this change necessary? This `__init__` is never called in case of C++ descriptors (since `__new__` returned a different object)

Comment (belvedere-trading) [inline python/google/protobuf/descriptor.py]: That's a good point. I added this originally because there was a similar check later in `__init__` (previously line 850). If there's no chance of the `cpp` implementation type making it to this code path, I can remove those checks and their related actions from `__init__` entirely.

**Comment by belvedere-trading** [comment]
This seems good to go on our end. Please let me know if you'd like me to remove the `cpp` checks from `__init__` as part of this PR as well.

**Comment by belvedere-trading** [comment]
Also, is there action required on our part to get the CLA check working?

**Comment by xfxyjwf** [comment]
@belvedere-trading How did you sign the CLA? Also is it a corp CLA, or individual CLA?

**Comment by belvedere-trading** [comment]
We signed it as a corporate CLA - Belvedere Trading.

**Comment by xfxyjwf** [comment]
@belvedere-trading I can verify that Belvedere Trading has signed Google corp CLA, but the CLA checker still fails on your github username. Have you added the email address associated with your github account to the google groups used for corporate CLA?

**Comment by belvedere-trading** [comment]
I just added my own email (the one associated with the two commits, jkaye@belvederetrading.com) to the Google group. Sorry, didn't realize I had to do that as well!

**Comment by xfxyjwf** [comment]
Thanks. The CLA is good now.

I'll merge after @amauryfa approves this pull request.

**Comment by belvedere-trading** [comment]
@xfxyjwf Any idea when we might be able to expect review on this? We need this for an internal build; we can deploy from our fork, but we'd prefer to stick to the official repository if possible.

**Comment by xfxyjwf** [comment]
ping @amauryfa

**Comment by belvedere-trading** [comment]
@amauryfa Any updates?

**Comment by belvedere-trading** [comment]
Is there anyone else who can review this?

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by xfxyjwf** [comment]
@haberman Can you help review this change?

File: python/google/protobuf/descriptor.py (hunk 1 above)
Comment (belvedere-trading) [inline python/google/protobuf/descriptor.py]: This code was here previously, so I was waiting on some confirmation that was the case. See previous line 850. I can remove if that's the consensus here.

Comment (belvedere-trading) [inline python/google/protobuf/descriptor.py]: Has someone looked into whether or not you'd like that code to be deleted?

**Comment by amauryfa** [comment]
Hi, sorry for the long delay.
First one remark: we usually build a FileDescriptorProto and call
DescriptorPool.Add(), it is easier. IMO the Descriptors should not be
created directly (this is how the c++ implementation works)
Then, I don't understand the last diff. How is descriptor_pool.Default()
different from _message.default_pool?

**Comment by belvedere-trading** [comment]
That entire portion of the code is unchanged from how it was previously (save it being moved to slightly later in the method); I'm not familiar with the details or why that was done in the first place. I can certainly change it if you think there's a better way, but my original intent with this PR was to address the issue of the pool argument not being supplied - I did not want to make unnecessary modifications.

**Comment by belvedere-trading** [comment]
How should I proceed with this?

**Comment by xfxyjwf** [comment]
@anandolee Jie, can you help review this pull request?

**Comment by belvedere-trading** [comment]
Bump...

**Comment by timdime** [comment]
Can an admin verify this patch?

File: python/google/protobuf/descriptor.py (hunk 1 above)
Comment (anandolee) [reply python/google/protobuf/descriptor.py]: Yes, the check here and original line 150 should be deleted.

File: python/google/protobuf/descriptor.py [hunk 2]
```diff
@@ -819,9 +819,9 @@ def __new__(cls, name, package, options=None, serialized_pb=None,
       # FileDescriptor() is called from various places, not only from generated
       # files, to register dynamic proto files and messages.
       if serialized_pb:
-        # TODO(amauryfa): use the pool passed as argument. This will work only
-        # for C++-implemented DescriptorPools.
-        return _message.default_pool.AddSerializedFile(serialized_pb)
+        # Does this have to be changed to support non-C++ backends?
+        _pool = pool or _message.default_pool
```
Comment (anandolee) [inline python/google/protobuf/descriptor.py]: Can you please add test for it?

**Comment by anandolee** [comment]
Sorry we missed the PR for a long time. Is it still an issue for your project?

**Comment by belvedere-trading** [comment]
Hi. This is still important for us, yes. We ended up forking the project a little more than a year ago and building our own packages to get this fix into our environment.

I can make the requested changes later this week or early next in hopes that we can get them merged.

**Comment by anandolee** [comment]
Thanks. This is one of pure python and cpp extension diffs that need to be fixed. Just test is required for such change.

**Comment by xfxyjwf** [comment]
@belvedere-trading Can you help add a test case for the change?

**Comment by belvedere-trading** [comment]
I plan on taking a look at this tomorrow.

**Comment by belvedere-trading** [comment]
Do your tests run using the C++ backend? I am seeing this:

```
(protobuf) jkaye-macbook:python jkaye$ PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=cpp python setup.py test
running test
running egg_info
writing requirements to protobuf.egg-info/requires.txt
writing protobuf.egg-info/PKG-INFO
writing namespace_packages to protobuf.egg-info/namespace_packages.txt
writing top-level names to protobuf.egg-info/top_level.txt
writing dependency_links to protobuf.egg-info/dependency_links.txt
reading manifest file 'protobuf.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files found matching 'google/protobuf/internal/*.proto'
warning: no previously-included files matching '*_test.py' found under directory 'google'
warning: no previously-included files matching '*_test.proto' found under directory 'google'
warning: no previously-included files matching '*.dll' found anywhere in distribution
warning: no previously-included files matching '*.pyc' found anywhere in distribution
warning: no previously-included files matching '*.pyo' found anywhere in distribution
warning: no previously-included files matching '*.so' found anywhere in distribution
writing manifest file 'protobuf.egg-info/SOURCES.txt'
running build_ext
Traceback (most recent call last):
  File "setup.py", line 252, in <module>
    ext_modules=ext_module_list,
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/Users/jkaye/.virtualenvs/protobuf/lib/python2.7/site-packages/setuptools/command/test.py", line 210, in run
    self.run_tests()
  File "/Users/jkaye/.virtualenvs/protobuf/lib/python2.7/site-packages/setuptools/command/test.py", line 231, in run_tests
    testRunner=self._resolve_as_ep(self.test_runner),
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/unittest/main.py", line 149, in parseArgs
    self.createTests()
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/unittest/main.py", line 158, in createTests
    self.module)
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/unittest/loader.py", line 130, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/unittest/loader.py", line 103, in loadTestsFromName
    return self.loadTestsFromModule(obj)
  File "/Users/jkaye/.virtualenvs/protobuf/lib/python2.7/site-packages/setuptools/command/test.py", line 42, in loadTestsFromModule
    tests.append(self.loadTestsFromName(submodule))
  File "/Users/jkaye/.pyenv/versions/2.7.13/lib/python2.7/unittest/loader.py", line 100, in loadTestsFromName
    parent, obj = obj, getattr(obj, part)
AttributeError: 'module' object has no attribute 'type_checkers'
```

**Comment by anandolee** [comment]
Our tests failed for cpp extension on your change with another error:
https://travis-ci.org/google/protobuf/builds/171196629

**Comment by anandolee** [comment]
Look like our python tests started fail 8 days ago... I will check why

**Comment by belvedere-trading** [comment]
That's fine. Looking at the Kokoro output, it seems that the tests are able to run with `python setup.py test`, but that same command gives me the exception I provided above. I'm running on 2.7.13 with the latest version of all dependencies if that helps.

**Comment by belvedere-trading** [comment]
I'm not talking about on Travis.. I cannot get these tests to run in any Python environment at all. The stack trace I posted above is an attempt to run the tests locally.

**Comment by anandolee** [comment]
Are you following the steps in README:
https://github.com/google/protobuf/blob/master/python/README.md

**Comment by anandolee** [comment]
I am closing it for clean up. Feel free to reopen if still need
